### PR TITLE
Add missing dependencies for: _khash_primitive_helper

### LINKED ISSUE #55615: BUILD: Building with samurai fails while ninja works.
### Installation check

- [X] I have read the [installation guide](https://pandas.pydata.org/pandas-docs/stable/getting_started/install.html#installing-pandas).

### Platform

I'm building with wolfi: https://github.com/wolfi-dev/os

### Installation Method

Built from source

### pandas Version

2.1.0, 2.1.1, 2.2.0.dev0

### Python Version

3.12

### Installation Logs

<details>
I was having a hard time building from source, and it would fail at some point with this:

```
/home/build/pandas/_libs/khash.pxd:129:0: 'khash_for_primitive_helper.pxi' not found
```

Which lead me to find this:
https://github.com/pandas-dev/pandas/issues/51875

But then I started looking at your CI pipelines to see how you're building them and trying to replicate, and noticed the build order is different. What I see in the CI pipelines is that the first things being built are:
```
    [1/151] Generating pandas/_libs/khash_primitive_helper_pxi with a custom command
    [2/151] Generating pandas/_libs/algos_common_helper_pxi with a custom command
    [3/151] Generating pandas/_libs/algos_take_helper_pxi with a custom command
    [4/151] Generating pandas/_libs/hashtable_class_helper_pxi with a custom command
    [5/151] Generating pandas/_libs/index_class_helper_pxi with a custom command
    [6/151] Generating pandas/_libs/hashtable_func_helper_pxi with a custom command
    [7/151] Generating pandas/__init__.py with a custom command
    [8/151] Generating pandas/_libs/intervaltree_helper_pxi with a custom command
    [9/151] Generating pandas/_libs/sparse_op_helper_pxi with a custom command
```

But when I was trying to build it, it was looking like this:
```
⚠️  aarch64   |   [1/152] Compiling Cython source /home/build/pandas/_libs/window/indexers.pyx
⚠️  aarch64   |   [2/152] Compiling Cython source /home/build/pandas/_libs/window/aggregations.pyx
⚠️  aarch64   |   [3/152] Compiling Cython source /home/build/pandas/_libs/writers.pyx
⚠️  aarch64   |   [4/152] Compiling Cython source /home/build/pandas/_libs/testing.pyx
⚠️  aarch64   |   [5/152] Compiling Cython source /home/build/pandas/_libs/tslib.pyx
⚠️  aarch64   |   [6/152] Generating pandas/_libs/sparse_op_helper_pxi with a custom command
⚠️  aarch64   |   [7/152] Compiling Cython source /home/build/pandas/_libs/byteswap.pyx
⚠️  aarch64   |   [8/152] Compiling Cython source /home/build/pandas/_libs/sas.pyx
```

And I dug into the differences for awhile more, and the one difference was that I was using [samurai](https://github.com/michaelforney/samurai) instead of `ninja` for the builds, and just by swapping to ninja I was able to make the build work. So, that's great. However, just wanted to bring this to your attention in case it may be something that might be masking a real issue (not saying there is one 😆 )

https://github.com/michaelforney/samurai#differences-from-ninja

```
samurai schedules jobs using a stack, so the last scheduled job is the first to execute, while ninja schedules jobs based on the pointer value of the edge structure (they are stored in a std::set<Edge*>), so the first to execute depends on the address returned by malloc. This may result in build failures due to insufficiently specified dependencies in the project's build system.
```

So, I don't yet know enough about how things are all configured for dependencies, and if that may be a red herring, but like I said just wanted to share that in case it tingles somebody's spidey sense.

</details>

### PR DESCRIPTION
**Pull request by vaikas**
- [ x ] closes #55615 - Add missing dependencies and one can now use `samurai` to build.

Thanks to @kaniini for helping me find where to add the dependencies.

### CODE REVIEW

**Comment by WillAyd** [comment]
Might be overlooking it but why do these modules need this as a source? Can you provide the original error you were getting?

**Comment by vaikas** [comment]
Thank you! Yeah, the original errors are in the issue under `Installation Logs` under `Details`, maybe I put them in the wrong section 🤣 . I did also dig deeper that has more details in this issue:
https://github.com/wolfi-dev/os/issues/7334

**Comment by kaniini** [comment]
When building with samurai, the dependency rules are sorted after loading, while the original implementation of ninja orders the dependency rules based on when they were malloc'd.  This can result in underdefined dependencies appearing to work simply as a side effect of when the dependency rule was evaluated.

In this case, both `arrays.pyx` and `index.pyx` are pulling in `pandas._libs` via an `import` (*not* `cimport`) statement, which results in `khash.pyx` being evaluated as part of the cython generation process.  But `khash.pyx` requires generated code, so it needs to be generated before these modules so they can safely pull in `pandas._libs`.

**Comment by WillAyd** [comment]
Do these need to be listed in `sources` or can you add them as `deps`? I see a few other modules already use deps and that seems like a more natural fit

**Comment by kaniini** [comment]
Yes, probably.  Testing it now!

**Comment by kaniini** [comment]
Sadly, Meson reports when using `deps`:

```
../pandas/_libs/meson.build:106:7: ERROR: Argument is of an unacceptable type 'CustomTarget'.
```

So I think we need to keep it as a source due to limitations of meson.

**Comment by WillAyd** [comment]
Did you declare that dependency against `_khash_primitive_helper` or `_khash_primitive_helper_dep`? Looks like the latter is used in a few places already so would be surprised if that didn't work

**Comment by vaikas** [comment]
Ok, phew that was easy 😆  PTAL, looking better?

**Review by WillAyd — APPROVED** [review]
Lgtm. Thanks for fixing up algos too. 

@mroeschke @lithomas1 any objections?

**Review by mroeschke — APPROVED** [review]
No objections on my side

File: pandas/_libs/meson.build [hunk 1]
```diff
@@ -61,12 +61,12 @@ subdir('tslibs')
 libs_sources = {
     # Dict of extension name -> dict of {sources, include_dirs, and deps}
     # numpy include dir is implicitly included
-    'algos': {'sources': ['algos.pyx', _algos_common_helper, _algos_take_helper, _khash_primitive_helper]},
-    'arrays': {'sources': ['arrays.pyx']},
+    'algos': {'sources': ['algos.pyx', _algos_common_helper, _algos_take_helper], 'deps': _khash_primitive_helper_dep},
+    'arrays': {'sources': ['arrays.pyx'], 'deps': _khash_primitive_helper_dep},
```
Comment (jbrockmendel) [inline pandas/_libs/meson.build:65]: seems weird that this would be needed here

Comment (lithomas1) [reply pandas/_libs/meson.build:65]: Agreed with Brock, doesn't arrays not cimport anything from pandas?

Comment (vaikas) [reply pandas/_libs/meson.build]: Right you are! 👍 I reckon it hit that in 'earlier' tries, and the proper fix was in index? 🤷

Comment (vaikas) [reply pandas/_libs/meson.build]: Well, it does seem to build, but breaks the unit tests?

Comment (lithomas1) [reply pandas/_libs/meson.build]: No, that's unrelated. 

#55804 should fix it.

**Review by lithomas1 — APPROVED** [review]
LGTM.

**Comment by vaikas** [comment]
I don't think the failing test is related to this PR looking at other PRs that are failing similarly?

**Comment by mroeschke** [comment]
Thanks @vaikas

**Pull request merged**